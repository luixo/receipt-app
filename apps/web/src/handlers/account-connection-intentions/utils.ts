import { TRPCError } from "@trpc/server";
import type { z } from "zod";

import { ACCOUNT_CONNECTIONS_INTENTIONS } from "~db/consts";
import type { AccountsId, UsersId } from "~db/models";
import type { Database } from "~db/types";
import type { emailSchema } from "~web/handlers/validation";

export const addConnectionIntention = async (
	database: Database,
	fromAccountId: AccountsId,
	user: { name: string },
	toEmail: z.infer<typeof emailSchema>,
	asUserId: UsersId,
) => {
	const targetAccount = await database
		.selectFrom("accounts")
		.select(["id", "avatarUrl"])
		.where("email", "=", toEmail.lowercase)
		.executeTakeFirst();
	if (!targetAccount) {
		throw new TRPCError({
			code: "NOT_FOUND",
			message: `Account with email "${toEmail.original}" does not exist.`,
		});
	}
	const connectedUser = await database
		.selectFrom("accounts")
		.where("accounts.email", "=", toEmail.lowercase)
		.innerJoin("users", (qb) =>
			qb
				.onRef("users.connectedAccountId", "=", "accounts.id")
				.on("users.ownerAccountId", "=", fromAccountId),
		)
		.select("users.name")
		.executeTakeFirst();
	if (connectedUser) {
		throw new TRPCError({
			code: "CONFLICT",
			message: `Account with email "${toEmail.original}" is already connected to user "${connectedUser.name}".`,
		});
	}
	const viceVersaIntention = await database
		.selectFrom("accountConnectionsIntentions")
		.select("userId")
		.where((eb) =>
			eb.and({
				accountId: targetAccount.id,
				targetAccountId: fromAccountId,
			}),
		)
		.executeTakeFirst();
	if (viceVersaIntention) {
		await database.transaction().execute(async (tx) => {
			await tx
				.updateTable("users")
				.set({ connectedAccountId: fromAccountId })
				.where((eb) =>
					eb.and({
						ownerAccountId: targetAccount.id,
						id: viceVersaIntention.userId,
					}),
				)
				.executeTakeFirst();
			await tx
				.updateTable("users")
				.set({ connectedAccountId: targetAccount.id })
				.where((eb) =>
					eb.and({
						ownerAccountId: fromAccountId,
						id: asUserId,
					}),
				)
				.executeTakeFirst();
			await tx
				.deleteFrom("accountConnectionsIntentions")
				.where((eb) =>
					eb.and({
						accountId: targetAccount.id,
						targetAccountId: fromAccountId,
					}),
				)
				.executeTakeFirst();
		});
		return {
			account: {
				id: targetAccount.id,
				email: toEmail.lowercase,
				avatarUrl: targetAccount.avatarUrl || undefined,
			},
			connected: true,
			user: {
				name: user.name,
			},
		};
	}
	try {
		await database
			.insertInto("accountConnectionsIntentions")
			.values({
				accountId: fromAccountId,
				targetAccountId: targetAccount.id,
				userId: asUserId,
				createdAt: new Date(),
			})
			.executeTakeFirst();
		return {
			account: {
				id: targetAccount.id,
				email: toEmail.lowercase,
				avatarUrl: targetAccount.avatarUrl || undefined,
			},
			connected: false,
			user: {
				name: user.name,
			},
		};
	} catch (e) {
		// Could be like `duplicate key value violates unique constraint "..."`
		const message = String(e);
		if (
			message.includes(ACCOUNT_CONNECTIONS_INTENTIONS.CONSTRAINTS.ACCOUNT_PAIR)
		) {
			const existingIntention = await database
				.selectFrom("accountConnectionsIntentions")
				.where((eb) =>
					eb.and({
						accountId: fromAccountId,
						targetAccountId: targetAccount.id,
					}),
				)
				.select(["userId"])
				.executeTakeFirstOrThrow();
			const existingUser = await database
				.selectFrom("users")
				.select("name")
				.where("id", "=", existingIntention.userId)
				.executeTakeFirstOrThrow();
			throw new TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to "${toEmail.original}" as user "${existingUser.name}".`,
			});
		}
		if (
			message.includes(ACCOUNT_CONNECTIONS_INTENTIONS.CONSTRAINTS.USER_PAIR)
		) {
			throw new TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to user "${user.name}".`,
			});
			// This is probably a c8 bug
			/* c8 ignore next */
		}
		/* c8 ignore next 2 */
		throw e;
	}
};
