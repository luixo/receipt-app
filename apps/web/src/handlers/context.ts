import { type H3Event, createEvent } from "@tanstack/react-start/server";
import type { inferProcedureBuilderResolverOptions } from "@trpc/server";
import type { TRPCRequestInfo } from "@trpc/server/http";
import type { IncomingMessage, ServerResponse } from "node:http";

import type { Database } from "~db/database";
import type { TestContext } from "~tests/backend/utils/test";
import type { authProcedure } from "~web/handlers/trpc";
import type { CacheDbOptions } from "~web/providers/cache-db";
import type { EmailOptions } from "~web/providers/email";
import type { ExchangeRateOptions } from "~web/providers/exchange-rate";
import type { Logger } from "~web/providers/logger";
import type { S3Options } from "~web/providers/s3";

type TestContextPicks = Pick<TestContext, "getSalt" | "getUuid"> & {
	database: Database;
	logger: Logger;
	emailOptions: EmailOptions;
	cacheDbOptions: CacheDbOptions;
	exchangeRateOptions: ExchangeRateOptions;
	s3Options: S3Options;
};

export type NetContext = {
	event: H3Event;
	info: TRPCRequestInfo;
};

export type UnauthorizedContext = NetContext & TestContextPicks;

export type AuthorizedContext = inferProcedureBuilderResolverOptions<
	typeof authProcedure
>["ctx"];

export const createContext = (
	{
		req,
		res,
		info,
	}: { req: IncomingMessage; res: ServerResponse; info: TRPCRequestInfo },
	testContext: TestContextPicks,
): UnauthorizedContext => ({
	info,
	event: createEvent(req, res),
	...testContext,
});
