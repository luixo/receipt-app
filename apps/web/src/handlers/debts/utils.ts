import type { Database, DebtsId, SimpleInsertObject } from "~db";
import { nonNullishGuard } from "~utils";

export const upsertAutoAcceptedDebts = async (
	database: Database,
	debts: (SimpleInsertObject<"debts"> & { isNew: boolean })[],
) => {
	const fetchedDebts = await database
		.selectFrom("debts")
		.where((eb) =>
			eb.or(
				debts.map((debt) =>
					!debt.isNew
						? eb.and({
								id: debt.id,
								ownerAccountId: debt.ownerAccountId,
						  })
						: debt.receiptId
						? eb.and({
								ownerAccountId: debt.ownerAccountId,
								userId: debt.userId,
								receiptId: debt.receiptId,
						  })
						: eb("debts.ownerAccountId", "is", null),
				),
			),
		)

		.select([
			"debts.id",
			"debts.ownerAccountId",
			"debts.userId",
			"debts.receiptId",
		])
		.execute();
	const counterpartyDebts = debts.map((debt) =>
		fetchedDebts.find((fetchedDebt) => {
			if (debt.isNew) {
				if (!debt.receiptId || !fetchedDebt.receiptId) {
					return false;
				}
				return (
					fetchedDebt.ownerAccountId === debt.ownerAccountId &&
					fetchedDebt.userId === debt.userId &&
					fetchedDebt.receiptId === debt.receiptId
				);
			}
			return (
				fetchedDebt.ownerAccountId === debt.ownerAccountId &&
				fetchedDebt.id === debt.id
			);
		}),
	);
	const existentDebts = debts
		.map((nextDebt, index) =>
			counterpartyDebts[index]
				? ([nextDebt, counterpartyDebts[index]!] as const)
				: null,
		)
		.filter(nonNullishGuard);
	const nonExistentDebts = debts
		.map((nextDebt, index) => (counterpartyDebts[index] ? null : nextDebt))
		.filter(nonNullishGuard);
	const [newDebts, ...updatedDebts] = await Promise.all([
		nonExistentDebts.length === 0
			? ([] as { id: DebtsId }[])
			: database
					.insertInto("debts")
					.values(nonExistentDebts.map(({ isNew, ...debt }) => debt))
					.returning(["debts.id"])
					.execute(),
		...existentDebts.map(([nextDebt, currentDebt]) =>
			database
				.updateTable("debts")
				.set({
					amount: nextDebt.amount,
					timestamp: nextDebt.timestamp,
					currencyCode: nextDebt.currencyCode,
					receiptId: nextDebt.receiptId,
					lockedTimestamp: nextDebt.lockedTimestamp,
				})
				.where((eb) =>
					eb.and({
						id: currentDebt.id,
						ownerAccountId: currentDebt.ownerAccountId,
						userId: currentDebt.userId,
					}),
				)
				.returning(["debts.id", "debts.receiptId"])
				.executeTakeFirst(),
		),
	]);
	return { newDebts, updatedDebts };
};
