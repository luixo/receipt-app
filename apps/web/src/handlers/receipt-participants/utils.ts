import type { Selection } from "kysely";

import type { ReceiptsId, UsersId } from "~db/models";
import type {
	Database,
	ReceiptsDatabase,
	ReceiptsSelectExpression,
} from "~db/types";

export const getReceiptParticipant = <
	SE extends ReceiptsSelectExpression<"receiptParticipants">,
>(
	database: Database,
	userId: UsersId,
	receiptId: ReceiptsId,
	selectExpression: SE[],
): Promise<
	Selection<ReceiptsDatabase, "receiptParticipants", SE> | undefined
> =>
	database
		.selectFrom("receiptParticipants")
		.where((eb) => eb.and({ receiptId, userId }))
		.select(selectExpression)
		.limit(1)
		.executeTakeFirst();
