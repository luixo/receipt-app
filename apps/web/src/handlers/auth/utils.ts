import { TRPCError } from "@trpc/server";

import { DAY } from "~utils";
import type { AccountsId, SessionsSessionId } from "~web/db/models";
import { generateConfirmEmailEmail } from "~web/email/utils";
import type { UnauthorizedContext } from "~web/handlers/context";
import { getEmailClient } from "~web/providers/email";

// How long a session should last
export const SESSION_EXPIRATION_DURATION = 30 * DAY;
// How long session should be intact before we update it
export const SESSION_SHOULD_UPDATE_EVERY = 2 * DAY;

export const getExpirationDate = () =>
	new Date(Date.now() + SESSION_EXPIRATION_DURATION);

export const createAuthorizationSession = async (
	ctx: UnauthorizedContext,
	accountId: AccountsId,
) => {
	const uuid: SessionsSessionId = ctx.getUuid();
	const expirationDate = getExpirationDate();
	await ctx.database
		.insertInto("sessions")
		.values({
			accountId,
			sessionId: uuid,
			expirationTimestamp: expirationDate,
		})
		.executeTakeFirst();
	return {
		authToken: uuid,
		expirationDate,
	};
};

export const sendVerificationEmail = async (
	ctx: UnauthorizedContext,
	email: string,
	token: string,
) => {
	if (!ctx.emailOptions.active) {
		throw new TRPCError({
			code: "FORBIDDEN",
			message: "Currently email resend is not supported",
		});
	}
	try {
		await getEmailClient(ctx).send({
			address: email,
			subject: "Confirm email in Receipt App",
			body: generateConfirmEmailEmail(ctx, token),
		});
	} catch (e) {
		throw new TRPCError({
			code: "INTERNAL_SERVER_ERROR",
			message: `Something went wrong: ${
				/* c8 ignore next */
				e instanceof Error ? e.message : String(e)
			}`,
		});
	}
};
