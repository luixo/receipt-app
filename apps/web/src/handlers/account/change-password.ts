import { TRPCError } from "@trpc/server";
import { z } from "zod";

import { passwordSchema } from "~app/utils/validation";
import { authProcedure } from "~web/handlers/trpc";
import { generatePasswordData, getHash } from "~web/utils/crypto";

export const procedure = authProcedure
	.input(
		z.strictObject({
			prevPassword: passwordSchema,
			password: passwordSchema,
		}),
	)
	.mutation(async ({ input, ctx }) => {
		const { database } = ctx;
		const account = await database
			.selectFrom("accounts")
			.select(["passwordHash", "passwordSalt"])
			.where("id", "=", ctx.auth.accountId)
			.executeTakeFirstOrThrow();
		const isPrevPasswordValid =
			getHash(input.prevPassword, account.passwordSalt) ===
			account.passwordHash;
		if (!isPrevPasswordValid) {
			throw new TRPCError({
				code: "UNAUTHORIZED",
				message: `Change password of account "${ctx.auth.email}" failed: password doesn't match.`,
			});
		}
		const passwordData = generatePasswordData(ctx, input.password);
		await database
			.updateTable("accounts")
			.set({
				passwordHash: passwordData.hash,
				passwordSalt: passwordData.salt,
			})
			.where("accounts.id", "=", ctx.auth.accountId)
			.executeTakeFirst();
	});
