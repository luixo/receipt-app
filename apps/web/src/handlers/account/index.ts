import { procedure as changeAvatar } from "./change-avatar";
import { procedure as changeName } from "./change-name";
import { procedure as changePassword } from "./change-password";
import { procedure as get } from "./get";
import { procedure as logout } from "./logout";
import { procedure as resendEmail } from "./resend-email";

export const router = {
	get,
	logout,
	changeName,
	changePassword,
	changeAvatar,
	resendEmail,
};
