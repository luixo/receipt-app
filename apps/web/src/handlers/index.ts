import { t } from "~web/handlers/trpc";

import { router as accountRouter } from "./account/index";
import { router as accountConnectionIntentionsRouter } from "./account-connection-intentions/index";
import { router as accountSettingsRouter } from "./account-settings/index";
import { router as adminRouter } from "./admin/index";
import { router as authRouter } from "./auth/index";
import { router as currencyRouter } from "./currency/index";
import { router as debtIntentionsRouter } from "./debt-intentions/index";
import { router as debtsRouter } from "./debts/index";
import { router as receiptItemConsumersRouter } from "./receipt-item-consumers/index";
import { router as receiptItemPayersRouter } from "./receipt-item-payers/index";
import { router as receiptItemsRouter } from "./receipt-items/index";
import { router as receiptParticipantsRouter } from "./receipt-participants/index";
import { router as receiptsRouter } from "./receipts/index";
import { router as resetPasswordIntentionsRouter } from "./reset-password-intentions/index";
import { router as sessionsRouter } from "./sessions/index";
import { router as usersRouter } from "./users/index";
import { router as utilsRouter } from "./utils/index";

export const router = t.router({
	// No auth
	sessions: sessionsRouter,
	auth: authRouter,
	resetPasswordIntentions: resetPasswordIntentionsRouter,
	utils: utilsRouter,
	// Auth
	account: accountRouter,
	accountSettings: accountSettingsRouter,
	receipts: receiptsRouter,
	receiptItems: receiptItemsRouter,
	users: usersRouter,
	receiptParticipants: receiptParticipantsRouter,
	currency: currencyRouter,
	receiptItemConsumers: receiptItemConsumersRouter,
	receiptItemPayers: receiptItemPayersRouter,
	accountConnectionIntentions: accountConnectionIntentionsRouter,
	debts: debtsRouter,
	debtIntentions: debtIntentionsRouter,
	// Admin
	admin: adminRouter,
});
