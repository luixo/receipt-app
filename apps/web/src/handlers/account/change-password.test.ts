import { faker } from "@faker-js/faker";
import { describe } from "vitest";

import {
	MAX_PASSWORD_LENGTH,
	MIN_PASSWORD_LENGTH,
} from "~app/utils/validation";
import { createAuthContext } from "~tests/backend/utils/context";
import { insertAccountWithSession } from "~tests/backend/utils/data";
import {
	expectDatabaseDiffSnapshot,
	expectTRPCError,
	expectUnauthorizedError,
} from "~tests/backend/utils/expect";
import { test } from "~tests/backend/utils/test";
import { t } from "~web/handlers/trpc";

import { procedure } from "./change-password";

const createCaller = t.createCallerFactory(t.router({ procedure }));

describe("account.changePassword", () => {
	describe("input verification", () => {
		expectUnauthorizedError((context) =>
			createCaller(context).procedure({ prevPassword: "", password: "" }),
		);

		const types = ["password", "prevPassword"] as const;
		types.forEach((type) => {
			const otherType = types.find((lookupType) => lookupType !== type);
			describe(type, () => {
				test("minimal length", async ({ ctx }) => {
					const { sessionId } = await insertAccountWithSession(ctx);
					const caller = createCaller(await createAuthContext(ctx, sessionId));
					await expectTRPCError(
						() =>
							caller.procedure({
								[type as "password"]: "a".repeat(MIN_PASSWORD_LENGTH - 1),
								[otherType as "prevPassword"]: "a".repeat(MIN_PASSWORD_LENGTH),
							}),
						"BAD_REQUEST",
						`Zod error\n\nAt "${type}": Minimal length for password is ${MIN_PASSWORD_LENGTH}`,
					);
				});

				test("maximum length", async ({ ctx }) => {
					const { sessionId } = await insertAccountWithSession(ctx);
					const caller = createCaller(await createAuthContext(ctx, sessionId));
					await expectTRPCError(
						() =>
							caller.procedure({
								[type as "password"]: "a".repeat(MAX_PASSWORD_LENGTH + 1),
								[otherType as "prevPassword"]: "a".repeat(MAX_PASSWORD_LENGTH),
							}),
						"BAD_REQUEST",
						`Zod error\n\nAt "${type}": Maximum length for password is ${MAX_PASSWORD_LENGTH}`,
					);
				});
			});
		});

		test("previous password doesn't match", async ({ ctx }) => {
			const currentPassword = faker.internet.password();
			const nextPassword = faker.internet.password();
			const {
				sessionId,
				account: { email },
			} = await insertAccountWithSession(ctx, {
				account: { password: currentPassword },
			});
			const caller = createCaller(await createAuthContext(ctx, sessionId));
			await expectTRPCError(
				() =>
					caller.procedure({
						password: nextPassword,
						prevPassword: `not_${currentPassword}`,
					}),
				"UNAUTHORIZED",
				`Change password of account "${email}" failed: password doesn't match.`,
			);
		});
	});

	describe("functionality", () => {
		test("password changes", async ({ ctx }) => {
			// Verifying other accounts are not affected
			await insertAccountWithSession(ctx);
			const currentPassword = faker.internet.password();
			const nextPassword = faker.internet.password();
			const { sessionId } = await insertAccountWithSession(ctx, {
				account: { password: currentPassword },
			});
			const caller = createCaller(await createAuthContext(ctx, sessionId));

			await expectDatabaseDiffSnapshot(ctx, () =>
				caller.procedure({
					password: nextPassword,
					prevPassword: currentPassword,
				}),
			);
		});
	});
});
