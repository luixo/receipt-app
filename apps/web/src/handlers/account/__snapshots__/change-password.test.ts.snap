// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`account.changePassword > functionality > password changes 1`] = `
{
  "updated": {
    "accounts": {
      "id:6e27b1dd-174b-4962-8c94-bdb98b6eaf50": {
        "passwordHash": "68b28f8324343380ee65453523e1abae23514174aaf780c5aaf4df584440783a2c393fa10483d3ed9c3cccc5bda857ae86a477d7591647c34867fa511a7b6fc7",
        "passwordSalt": "b7a83c6b0f1c72ec0ea3878dced9fda1a9c0002ccd40ca1a747e2d9a5a702f30a2b70aa1bcaeda8b8fccc52ef38b3b98b19ec8e327ba812f34a7fdc79af505e2",
        "updatedAt": 2020-01-01T00:01:00.000Z,
      },
    },
  },
}
`;
